File: CommonCoreLegacy/src/main/java/com/skedgo/tripkit/common/model/Regions.java
package com.skedgo.tripkit.common.model;

import android.os.Build;

import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.Objects;
import java.util.Set;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public final class Regions {
    private Regions() {
    }

    public static boolean equals(@Nullable Region a, @Nullable Region b) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            return Objects.equals(a, b);
        } else {
            return a != null && b != null && a.equals(b);
        }
    }

    @NonNull
    public static Region createInterRegion(
        @NonNull Region departureRegion,
        @NonNull Region arrivalRegion) {
        return new InterRegion(departureRegion, arrivalRegion);
    }

    static final class InterRegion extends Region {
        InterRegion(@NonNull Region departureRegion, @NonNull Region arrivalRegion) {
            setName(departureRegion.getName() + "_" + arrivalRegion.getName());

            final ArrayList<String> departureModeIds = departureRegion.getTransportModeIds();
            final ArrayList<String> arrivalModeIds = arrivalRegion.getTransportModeIds();
            if (departureModeIds != null && arrivalModeIds != null) {
                final Set<String> unionModeIds = new LinkedHashSet<>(
                    departureModeIds.size() + arrivalModeIds.size() + 1
                );
                unionModeIds.add(TransportMode.ID_AIR);
                unionModeIds.addAll(departureModeIds);
                unionModeIds.addAll(arrivalModeIds);
                setTransportModeIds(new ArrayList<>(unionModeIds));
            } else if (departureModeIds != null) {
                final Set<String> unionModeIds = new LinkedHashSet<>(
                    departureModeIds.size() + 1
                );
                unionModeIds.add(TransportMode.ID_AIR);
                unionModeIds.addAll(departureModeIds);
                setTransportModeIds(new ArrayList<>(unionModeIds));
            } else if (arrivalModeIds != null) {
                final Set<String> unionModeIds = new LinkedHashSet<>(
                    arrivalModeIds.size() + 1
                );
                unionModeIds.add(TransportMode.ID_AIR);
                unionModeIds.addAll(arrivalModeIds);
                setTransportModeIds(new ArrayList<>(unionModeIds));
            }

            final ArrayList<String> departureUrls = departureRegion.getURLs();
            if (departureUrls != null) {
                setURLs(new ArrayList<>(departureUrls));
            }

            setTimezone(departureRegion.getTimezone());
            setEncodedPolyline(departureRegion.getEncodedPolyline());
        }
    }
}

File: CommonCoreLegacy/src/main/java/com/skedgo/tripkit/common/model/region/Regions.kt
package com.skedgo.tripkit.common.model.region

import android.os.Build.VERSION
import android.os.Build.VERSION_CODES
import com.skedgo.tripkit.common.model.TransportMode

object Regions {
    @JvmStatic
    fun equals(a: Region?, b: Region?): Boolean {
        return a == b
    }

    @JvmStatic
    fun createInterRegion(
        departureRegion: Region,
        arrivalRegion: Region
    ): Region {
        return InterRegion(departureRegion, arrivalRegion)
    }

    class InterRegion(departureRegion: Region, arrivalRegion: Region) : Region() {
        init {
            name = departureRegion.name + "_" + arrivalRegion.name

            val departureModeIds = departureRegion.transportModeIds
            val arrivalModeIds = arrivalRegion.transportModeIds
            if (departureModeIds != null && arrivalModeIds != null) {
                val unionModeIds: MutableSet<String> = LinkedHashSet(
                    departureModeIds.size + arrivalModeIds.size + 1
                )
                unionModeIds.add(TransportMode.ID_AIR)
                unionModeIds.addAll(departureModeIds)
                unionModeIds.addAll(arrivalModeIds)
                transportModeIds = ArrayList(unionModeIds)
            } else if (departureModeIds != null) {
                val unionModeIds: MutableSet<String> = LinkedHashSet(
                    departureModeIds.size + 1
                )
                unionModeIds.add(TransportMode.ID_AIR)
                unionModeIds.addAll(departureModeIds)
                transportModeIds = ArrayList(unionModeIds)
            } else if (arrivalModeIds != null) {
                val unionModeIds: MutableSet<String> = LinkedHashSet(
                    arrivalModeIds.size + 1
                )
                unionModeIds.add(TransportMode.ID_AIR)
                unionModeIds.addAll(arrivalModeIds)
                transportModeIds = ArrayList(unionModeIds)
            }

            val departureUrls = departureRegion.getURLs()
            if (departureUrls != null) {
                setURLs(ArrayList(departureUrls))
            }

            timezone = departureRegion.timezone
            encodedPolyline = departureRegion.encodedPolyline
        }
    }
}